content/post/formatting-with-style.md

```diff
@@ -0,0 +1,247 @@
+---
+author: m09
+date: 2018-10-12
+title: "Formatting. With Style."
+image: /post/formatting-with-style/teapot.jpg
+description: "Formatting code using existing styles."
+categories: ["technical", "MLonCode"]
+---
+
+Ensuring that a codebase is consistent in style is both hard and
+costly, yet it is extremely important for maintainability and to
+reduce technical debt. Why is it hard and costly?  There are many
+contributing factors but the main ones are certainly:
+
+- teams might not have set up a linting process early enough or at all
+- people handling the codebase might be changing on a regular basis
+
+To allow new or current team members to code in a consistent way, we
+therefore need to model existing style and use it to lint new code. At
+source{d}, we call that breakfast. Let's go!
+
+## Lookout
+
+This problem is one of the many painpoints we are currently tackling
+with [Lookout](https://github.com/src-d/lookout), our brand new and
+frankly awesome assisted code review framework.
+
+The purpose of Lookout is to bring assisted code review to anyone in
+an easy-to-setup, easy-to-use, easy-to-extend fashion. To achieve
+that, Lookout watches Github repos and triggers a set of analyzers
+when new code is pull-request-ed or pushed. Those analyzers are very
+easy to define (they are based on the gRPC toolsuite).
+
+In our case, when new code is pushed, we want to learn from a codebase
+to model its style as precisely as possible. When new code is
+pull-request-ed, we analyze it to detect any styling issue and report
+it for consideration during code review.
+
+## Unsupervised Learning
+
+Let's now dive into the Machine Learning side of things! In our first
+approach and to come up with a solution that completely addressed this
+problem of linting code given an existing codebase, we defined several
+keypoints to achieve:
+
+- only require the code of one repository to be train our model
+- output something interpretable (to be accountable and establish
+  trust with our users —a key point of Machine Learning in production)
+- come up with an correctable model: if a team is not happy with the
+  suggestions of the machine-learning based model, it should be able
+  to correct its suggestions
+
+To satisfy all those needs with a single approach, we turned to good
+old, battle-tested machine learning methods: language models and
+decision trees.
+
+### Language models
+
+To define the style of a repository, we summoned a tool used every day
```

Comment:
In order to define the style of a repository, we turned to the well-known Natural Language Processing subjects: language models.